iceoryx_hoofs/test/moduletests/test_polymorphic_handler.cpp

```diff
@@ -0,0 +1,249 @@
+constexpr uint32_t ALTERNATE_ID = 21;
+
+struct Default : public Interface
+{
+    uint32_t id() override
+    {
+        return DEFAULT_ID;
+    }
+};
+
+struct Alternate : public Interface
+{
+    uint32_t id() override
+    {
+        return ALTERNATE_ID;
+    }
+};
+
+template <typename T>
+using Guard = iox::design_pattern::StaticLifetimeGuard<T>;
+
+// should the handler instances be accessed, they will live at least as
+// long as the guard objects
+Guard<Default> defaultGuard;
+Guard<Alternate> alternateGuard;
+
+// will live at least as long as the corresponding guards
+Default& defaultHandler = Guard<Default>::instance();
+Alternate& alternateHandler = Guard<Alternate>::instance();
+
+struct Hooks
+{
+    // to check whether the arguments are used correctly
+    static void onSetAfterFinalize(Interface& currentInstance, Interface& newInstance)
+    {
+        currentInstance.value = currentInstance.id();
+        newInstance.value = newInstance.id();
+    }
+};
+
+using Handler = iox::design_pattern::PolymorphicHandler<Interface, Default, Hooks>;
+
+class PolymorphicHandler_test : public Test
+{
+  public:
+    void SetUp() override
+    {
+        Handler::reset();
+    }
+
+    void TearDown() override
+    {
+    }
+};
+
+TEST_F(PolymorphicHandler_test, handlerIsInitializedWithDefault)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "41bb4a5e-a916-4a6d-80c4-fed3a3d8d78b");
+    auto& handler = Handler::get();
+
+    EXPECT_EQ(handler.id(), DEFAULT_ID);
+}
+
+TEST_F(PolymorphicHandler_test, settingAlternateWorks)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "8b2f0cfe-f13c-4fa0-aa93-5ddd4f0904d1");
+    Handler::set(alternateGuard);
+    auto& handler = Handler::get();
+
+    EXPECT_EQ(handler.id(), ALTERNATE_ID);
+}
+
+TEST_F(PolymorphicHandler_test, alternatePointsToExternalMemory)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "85ce0e51-a1fe-490c-9012-7d539512ed38");
+    Handler::set(alternateGuard);
+
+    auto& handler = Handler::get();
+    auto* ptr = static_cast<Interface*>(&alternateHandler);
+
+    EXPECT_EQ(&handler, ptr);
+}
+
+TEST_F(PolymorphicHandler_test, explicitlySettingToDefaultWorks)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "32e4d808-c848-4bf9-b878-e163ca825539");
+    Handler::set(alternateGuard);
+    Handler::set(defaultGuard);
+
+    auto& handler = Handler::get();
+    auto* ptr = static_cast<Interface*>(&defaultHandler);
+
+    EXPECT_EQ(&handler, ptr);
+}
+
+TEST_F(PolymorphicHandler_test, resetToDefaultWorks)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "ef8a99da-22a6-497e-b2ec-bf72cc3ae943");
+    Handler::set(alternateGuard);
+    Handler::reset();
+
+    auto& handler = Handler::get();
+    EXPECT_EQ(handler.id(), DEFAULT_ID);
+}
+
+TEST_F(PolymorphicHandler_test, settingAfterFinalizeCallsHook)
+{
+    ::testing::Test::RecordProperty("TEST_ID", "171ac802-01b9-4e08-80a6-6f2defecaf6d");
+
+    auto& handler = Handler::get();
+    handler.reset();
+    alternateHandler.reset();
```

Comment:
Here `Interface::reset` is called, which sets `value` to 0, correct?

Reply:
Yes. This is to gain confidence (cannot ensure 100%) that the hook sets the values as expected below (they could otherwise be correct by coincidence etc.)